test_contextual_lenses/contextual_lenses.py

```diff
@@ -0,0 +1,128 @@
+"""Contextual lenses."""
+"""Creates sequence length independent representation of embedded sequences."""
+"""Original paper: https://arxiv.org/pdf/2002.08866.pdf."""
+
+
+import flax
+from flax import nn
+
+import jax
+import jax.nn
+import jax.numpy as jnp
+
+import numpy as np
+
+
+def max_pool(x, padding_mask=None, **reduce_fn_kwargs):
+  """Apply padding, take maximum over sequence length axis."""
+
+  if padding_mask is not None:
+    x = x * padding_mask
+    neg_mask = - 999999. * (1 - padding_mask)
+    x = x + neg_mask
+   
+  rep = jnp.max(x, axis=-2)
+   
+  return rep
+
+
+def mean_pool(x, padding_mask=None, **reduce_fn_kwargs):
+  """Apply padding, take mean over sequence length axis."""
+
+  if padding_mask is not None:
+    x = x * padding_mask
+    rep = jnp.sum(x, axis=-2) / jnp.sum(padding_mask, axis=-2)
+  else:
+    rep = jnp.mean(x, axis=-2)
+
+  return rep
+
+
+def linear_max_pool(x, padding_mask=None, **reduce_fn_kwargs):
+  """Apply linear transformation + ReLU, apply padding,
+     take maximum over sequence length."""
+  
+  rep_size = list(reduce_fn_kwargs.values())[0]
+
+  x = nn.Dense(
+        x,
+        rep_size,
+        kernel_init=nn.initializers.xavier_uniform(),
+        bias_init=nn.initializers.normal(stddev=1e-6))
+  
+  x = nn.relu(x)
+  
+  rep = max_pool(x, padding_mask)
+  
+  return rep
+
+
+def linear_mean_pool(x, padding_mask=None, **reduce_fn_kwargs):
+  """Apply linear transformation + ReLU, apply padding,
+     take mean over sequence length."""
+  
+  rep_size = list(reduce_fn_kwargs.values())[0]
+  
+  x = nn.Dense(
+        x,
+        rep_size,
+        kernel_init=nn.initializers.xavier_uniform(),
+        bias_init=nn.initializers.normal(stddev=1e-6))
+  
+  x = nn.relu(x)
+  
+  rep = mean_pool(x, padding_mask)
+  
+  return rep
+
+
+class GatedConv(nn.Module):
+  """Gated Convolutional lens followed by max pooling,
+     see original paper for details."""
+
+  def apply(self, x, padding_mask=None, **reduce_fn_kwargs):
+    
+    rep_size, M_layers, M_features, M_kernel_sizes, zero_rep_size = list(reduce_fn_kwargs.values())
```

Comment:
this unpacking of kwargs is risky bc if the user provided them in a different order, you would get the wrong values. Essentially, you're using *args. You should switch to having all of these be named arguments, such that the function signature is apply(self, x, padding_mask=None, rep_size=10, M_layers=5, ...)
Here, I chose default values for these kwargs. Setting defaults can lead to bugs where you accidentally use default values when you meant to use non-default values. You could have the signature be apply(self, x, padding_mask, rep_size, M_layers, ...)